regex_helper.c

```diff
@@ -0,0 +1,86 @@
+#include <ctype.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <string.h>
+#include <syslog.h>
+#include <stdlib.h>
+
+#include "user_secret.h"
+#include "trace.h"
+
+#define min(a,b)            (((a) < (b)) ? (a) : (b))
+
+/* 
+ * Macros for user secret regex
+ * These are BRE regex, please refer to: https://en.wikibooks.org/wiki/Regular_Expressions/POSIX_Basic_Regular_Expressions
+ */
+#define USER_SECRET_REGEX_WHITE_SPACE              "[[:space:]]*"
+#define USER_SECRET_REGEX_SECRET                   "\\([^[:space:]]*\\)"
+
+/* Regex match group count, 2 because only have 1 subexpression for user secret */
+#define REGEX_MATCH_GROUP_COUNT      2
+
+/* The user secret mask */
+#define USER_SECRET_MASK                   '*'
+
+/* Replace user secret with regex */
+char * remove_user_secret_by_regex(const char* command, regex_t regex)
+{
+    regmatch_t pmatch[REGEX_MATCH_GROUP_COUNT];
+    if (regexec(&regex, command, REGEX_MATCH_GROUP_COUNT, pmatch, 0) == REG_NOMATCH) {
+        trace("User command not match.\n");
+        return NULL;
+    }
+
+    if (pmatch[1].rm_so < 0) {
+        trace("User secret not found.\n");
+        return NULL;
+    }
+
+    /* Found user secret between pmatch[1].rm_so to pmatch[1].rm_eo, replace it. */
+    trace("Found user secret between: %d -- %d\n", pmatch[1].rm_so, pmatch[1].rm_eo);
+    int result_buffer_size = strlen(command) + 1;
+    char * result_buffer = malloc(result_buffer_size);
+    snprintf(result_buffer, result_buffer_size, "%s", command);
+
+    /* Replace user secret with mask. */
+    int secret_start_pos = min(pmatch[1].rm_so, result_buffer_size - 1);
+    int secret_count = min(pmatch[1].rm_eo, result_buffer_size - 1) - secret_start_pos;
+    memset(result_buffer + secret_start_pos, USER_SECRET_MASK, secret_count);
+
+    return result_buffer;
+}
+
+/* Convert user secret setting to regex. */

```

Comment:
>secret setting [](http://example.com/codeflow?start=16&length=14)

`secret setting` is not easy to understand. I guess you mean `sudoers Cmnd`. ref: [https://www.sudo.ws/man/1.7.10/sudoers.man.html](https://www.sudo.ws/man/1.7.10/sudoers.man.html)

Reply:
Fixed, and rename to password.

Reply:
Should change password_settings to PASSWD_CMD to regex.

Reply:
Fixed.